pkg/erofs/erofs.go

```diff
@@ -0,0 +1,711 @@
+
+// IsDir indicates whether i represents a directory.
+func (i *Inode) IsDir() bool {
+	return i.fileType == FileTypeDir
+}
+
+// IsCharDev indicates whether i represents a character device.
+func (i *Inode) IsCharDev() bool {
+	return i.fileType == FileTypeCharDev
+}
+
+// IsBlockDev indicates whether i represents a block device.
+func (i *Inode) IsBlockDev() bool {
+	return i.fileType == FileTypeBlockDev
+}
+
+// IsFIFO indicates whether i represents a named pipe.
+func (i *Inode) IsFIFO() bool {
+	return i.fileType == FileTypeFIFO
+}
+
+// IsSocket indicates whether i represents a socket.
+func (i *Inode) IsSocket() bool {
+	return i.fileType == FileTypeSocket
+}
+
+// IsSymlink indicates whether i represents a symbolic link.
+func (i *Inode) IsSymlink() bool {
+	return i.fileType == FileTypeSymlink
+}
+
+// Nid returns the inode number.
+func (i *Inode) Nid() uint64 {
+	return i.nid
+}
+
+// Size returns the data size.
+func (i *Inode) Size() uint64 {
+	return i.size
+}
+
+// Nlink returns the number of hard links.
+func (i *Inode) Nlink() uint32 {
+	return i.nlink
+}
+
+// Mtime returns the time of last modification.
+func (i *Inode) Mtime() uint64 {
+	return i.mtime
+}
+
+// MtimeNsec returns the nano second part of Mtime.
+func (i *Inode) MtimeNsec() uint32 {
+	return i.mtimeNsec
+}
+
+// Mode returns the file type and permissions.
+func (i *Inode) Mode() uint16 {
+	return i.mode
+}
+
+// UID returns the user ID of the owner.
+func (i *Inode) UID() uint32 {
+	return i.uid
+}
+
+// GID returns the group ID of the owner.
+func (i *Inode) GID() uint32 {
+	return i.gid
+}
+
+// FD returns the host FD of underlaying image file.
+func (i *Inode) FD() int {
+	return int(i.image.src.Fd())
+}
+
+// DataOffset returns the offset of the file data of this inode in image file.
+// Only "flat file data without data inline (no extent)" data layout is supported.
+func (i *Inode) DataOffset() (uint64, error) {
+	if i.DataLayout() != InodeDataLayoutFlatPlain {
+		log.Warningf("Unsupported data layout 0x%x of inode (nid=%v)", i.DataLayout(), i.Nid())
+		return 0, linuxerr.ENOTSUP
+	}
+	return i.dataOff, nil
+}
+
+// Data returns the read-only file data of this inode. Only "flat file data
+// without data inline (no extent)" data layout is supported.
+func (i *Inode) Data() ([]byte, error) {
+	if i.DataLayout() != InodeDataLayoutFlatPlain {
+		log.Warningf("Unsupported data layout 0x%x of inode (nid=%v)", i.DataLayout(), i.Nid())
+		return nil, linuxerr.ENOTSUP
+	}
+	return i.image.BytesAt(i.dataOff, i.size)
+}
+
+// ensureChildMap populates the mappings of child names to inodes.
+//
+// Preconditions:
+//   - i.IsDir().
+func (i *Inode) ensureChildMap() error {
+	i.dirMu.Lock()
+	defer i.dirMu.Unlock()
+
+	if i.childMap != nil {
+		return nil
+	}
+
+	childMap := make(map[string]*Inode)
+
+	start := i.dataOff
+	if i.hasInlineData && i.blocks == 1 {
+		start = i.idataOff
+	}
+
+	blockSize := uint64(i.image.BlockSize())
+	blocks := i.blocks
+	for blocks > 0 {
```

Comment:
Not quite look into the implementation details, yet not sure where the on-disk formatlooks odd:

First, if you'd like to do dirent random access(like name binary search) in a block-based filesystem, you needs to make dirents arrange together instead of [(dirent + filename) | (dirent + filename)] approach;
Second, since it's a block-based filesystem, some padding is needed so that all metadata is block-aligned;
Third, introducing optional padding avoids an uncessary end-dirent.

Reply:
I see. The padding and block-aligned part made complete sense. I found it odd that the format was [dirent, dirent, dirent, name, name, name]. I thought that [dirent, name, dirent, name] made more sense (and is more straight forward to unmarshal). But like you said, it does kill random access in a block.

Good point. I see we can potentially do binary search with this format:
- In a block, use first dirent's NameOff to figure out how many dirents exist in that block.
- Query the mid dirent. Use its NameOff and the next Dirent's NameOff to read dirent name.
- Compare with search value. Continue binary search.

In EROFS, are dirents always sorted on disk?

Reply:
> In EROFS, are dirents always sorted on disk?

Yes, and we actually use it in kernel for random access via binary search :)

Reply:
> Could you please add comments across this function explaining each step and also the overall format.

Sure, will do.